### PR TITLE
Add AvgPool3D layer

### LINKED ISSUE #81: [Video] Add AveragePooling3D layer
We are missing some layers to support the export of models from Keras fully. One of them is the AveragePooling3D layer. A detailed description of the layer can be found [here](https://keras.io/api/layers/pooling_layers/average_pooling3d/).

- [ ] add layer class to the  [package](https://github.com/JetBrains/KotlinDL/tree/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling) using AvgPool2D as a [reference implementation](https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool2D.kt)
- [ ] add documentation
- [ ] add test
- [ ] add export function  (see [ModelSaver.kt](https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelSaver.kt)) 
- [ ] add import function (see [ModelLoader.kt](https://github.com/JetBrains/KotlinDL/blob/master/api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelLoader.kt))
- [ ] (optional) create an example

### PR DESCRIPTION
**Pull request by mkaze**
Resolves #81. This PR adds support for AvgPool3D layer.

- [x] Implement layer class
- [x] Add unit tests for layer implementation
- [x] Add support for importing models containing this layer from JSON config files
- [x] Add support for exporting models containing this layer to JSON config files

NOTE: as commented in the unit tests, currently it gives an error when `dataFormat = CHANNELS_FIRST`.

### CODE REVIEW

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ * @property [strides] The amount of shift for pooling window in each pooling step. If
+ * `null`, it will default to [poolSize].
+ * @property [padding] Padding strategy; can be either of [ConvPadding.VALID] which means no
+ * padding, or [ConvPadding.SAME] which means padding the input equally such that the output
+ * has the same dimension as the input.
+ * @property [dataFormat] Data format of input; can be either of [CHANNELS_LAST] or [CHANNELS_FIRST].
+ */
+public class AvgPool3D(
+    public val poolSize: IntArray = intArrayOf(2, 2, 2),
+    public val strides: IntArray? = null,
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt:34]: set the default value for the strides and convert both, strides and poolSize to the 5d array of longs

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    public val padding: ConvPadding = ConvPadding.VALID,
+    public val dataFormat: String = CHANNELS_LAST,
+    name: String = ""
+) : Layer(name) {
+
+    override val hasActivation: Boolean
+        get() = false
+    override val paramCount: Int
+        get() = 0
+    override val weights: Map<String, Array<*>>
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt:44]: merge with master and add setter

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+ * @property [padding] Padding strategy; can be either of [ConvPadding.VALID] which means no
+ * padding, or [ConvPadding.SAME] which means padding the input equally such that the output
+ * has the same dimension as the input.
+ * @property [dataFormat] Data format of input; can be either of [CHANNELS_LAST] or [CHANNELS_FIRST].
+ */
+public class AvgPool3D(
+    public val poolSize: IntArray = intArrayOf(2, 2, 2),
+    public val strides: IntArray? = null,
+    public val padding: ConvPadding = ConvPadding.VALID,
+    public val dataFormat: String = CHANNELS_LAST,
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt:36]: remove dataFormat

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        var dim2 = inputShape.size(axis1 + 1)
+        var dim3 = inputShape.size(axis1 + 2)
+        val strides1 = strides?.get(0) ?: poolSize[0]
+        val strides2 = strides?.get(1) ?: poolSize[1]
+        val strides3 = strides?.get(2) ?: poolSize[2]
+        dim1 = convOutputLength(dim1, poolSize[0], padding, strides1)
+        dim2 = convOutputLength(dim2, poolSize[1], padding, strides2)
+        dim3 = convOutputLength(dim3, poolSize[3], padding, strides3)
+
+        return if (dataFormat == CHANNELS_LAST) {
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt:79]: remove dataFormat

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+                expectedShape,
+                output.shape().toIntArray()
+            )
+
+            // Check output values are correct.
+            val actual = Array(input.size) {
+                Array(input[0].size) { Array(1) { Array(2) { FloatArray(1) } } }
+            }
+            output.tensor().copyTo(actual)
+            for (i in expected.indices) {
```
Comment (zaleslaw) [inline api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt:164]: remove this test due to removed dataFormat

**Review by zaleslaw — CHANGES_REQUESTED** [review]

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt (hunk 1 above)
Comment (mkaze) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt]: Conversion was done, and I have discussed the benefit of using `null` as the default value for strides.

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt (hunk 2 above)
Comment (mkaze) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt]: Done!

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt (hunk 3 above)
Comment (mkaze) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt]: Done!

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt (hunk 4 above)
Comment (mkaze) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt]: Done!

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt (hunk 5 above)
Comment (mkaze) [reply api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt]: Done!

**Comment by mkaze** [comment]
@zaleslaw I just merged it with master branch and updated it with recent/requested changes.

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+ * @property [poolSize] Size of the pooling window for each dimension of input.
+ * @property [strides] The amount of shift for pooling window in each pooling step. If
+ * `null`, it will default to [poolSize].
+ * @property [padding] Padding strategy; can be either of [ConvPadding.VALID] which means no
+ * padding, or [ConvPadding.SAME] which means padding the input equally such that the output
+ * has the same dimension as the input.
+ */
+public class AvgPool3D(
+    public val poolSize: LongArray = longArrayOf(1, 2, 2, 2, 1),
+    public val strides: LongArray? = null,
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt:30]: LongArray? -> LongArray with default value

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    override var weights: Map<String, Array<*>>
+        get() = emptyMap()
+        set(value) = assignWeights(value)
+
+    init {
+        require(poolSize.size == 5) {
+            "The poolSize should be an array of size 5."
+        }
+
+        require(strides == null || strides.size ==5) {
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/pooling/AvgPool3D.kt:48]: change the condition

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    @Test
+    fun withPaddingAndStride() {
+        val layer = AvgPool3D(strides = longArrayOf(1, 1, 1, 1, 1), padding = ConvPadding.SAME)
+        val expected = arrayOf(
+            arrayOf(
+                arrayOf(
+                    arrayOf(
+                        floatArrayOf(18.0f/8, 4.5f/8, 16.0f/8),
```
Comment (zaleslaw) [inline api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt:122]: did you take these values from Keras tests?

**Review by zaleslaw — CHANGES_REQUESTED** [review]

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt (hunk 8 above)
Comment (mkaze) [reply api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt:122]: No, I created the test data myself :)

**Comment by mkaze** [comment]
@zaleslaw I changed the strides type and default value per your request.

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt (hunk 8 above)
Comment (zaleslaw) [reply api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt:122]: Not bad:)
Sometimes I just copied test data from Keras; sometimes, if test data seems not appropriate, I use Keras like a test machine writing some simple test in Python to put some test data in inputs and get outputs. A few times, I ignored this method and just used tested primitive itself to get data, but it has a bug in implementation and, as a result, a wrong test:(( But I agree that test inputs/outputs for pooling could be calculated manually (just shared my experience)

Comment (mkaze) [reply api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/AvgPool3DTest.kt:122]: Yeah, actually I have created test data for all test cases I have written so far; but certainly it's not a sustainable approach and using another tested tool, like Keras or maybe even numpy, for generating test data seems to be a better and more reliable approach.

**Pull request merged**